src/screen.cpp

```diff
@@ -198,70 +202,77 @@ static bool is_csi_style_escape_seq(const wchar_t *code, size_t *resulting_lengt
-                              exit_alt_charset_mode};
-
-        for (size_t p = 0; p < sizeof esc2 / sizeof *esc2 && !found; p++) {
-            if (!esc2[p]) continue;
-            // Test both padded and unpadded version, just to be safe. Most versions of tparm don't
-            // actually seem to do anything these days.
-
-            size_t len = maxi(try_sequence(tparm(esc2[p]), code), try_sequence(esc2[p], code));
-            if (len) {
-                resulting_length = len;
-                found = true;
-            }
+    return false;
+}
+
+// Detect whether the escape sequence sets one of the terminal attributes that affects how text is
+// displayed other than the color.
+static bool is_visual_escape_seq(const wchar_t *code, size_t *resulting_length) {
+    if (!cur_term) return false;
+    char *const esc2[] = {enter_bold_mode,        exit_attribute_mode,   enter_underline_mode,
+                            exit_underline_mode,    enter_standout_mode,   exit_standout_mode,
+                            flash_screen,           enter_subscript_mode,  exit_subscript_mode,
+                            enter_superscript_mode, exit_superscript_mode, enter_blink_mode,
+                            enter_italics_mode,     exit_italics_mode,     enter_reverse_mode,
+                            enter_shadow_mode,      exit_shadow_mode,      enter_standout_mode,
+                            exit_standout_mode,     enter_secure_mode,     enter_dim_mode,
+                            enter_blink_mode,       enter_protected_mode,  enter_alt_charset_mode,
+                            exit_alt_charset_mode};
+
+    for (size_t p = 0; p < sizeof esc2 / sizeof *esc2; p++) {
+        if (!esc2[p]) continue;
+        // Test both padded and unpadded version, just to be safe. Most versions of tparm don't
+        // actually seem to do anything these days.
+        size_t esc_seq_len = maxi(try_sequence(tparm(esc2[p]), code), try_sequence(esc2[p], code));
+        if (esc_seq_len) {
+            *resulting_length = esc_seq_len;
+            return true;
         }
     }
 
-    if (!found) found = is_screen_name_escape_seq(code, &resulting_length);
-    if (!found) found = is_iterm2_escape_seq(code, &resulting_length);
-    if (!found) found = is_single_byte_escape_seq(code, &resulting_length);
-    if (!found) found = is_csi_style_escape_seq(code, &resulting_length);
-    if (!found) is_two_byte_escape_seq(code, &resulting_length);
+    return false;
+}
 
-    return resulting_length;
+/// Returns the number of characters in the escape code starting at 'code'. We only handle sequences
+/// that begin with \e. If it doesn't we return zero.
+size_t escape_code_length(const wchar_t *code) {
+    assert(code != NULL);
+    if (*code != L'\e') return 0;
+
+    size_t esc_seq_len = cached_esc_sequences.find_entry(code);
+    if (esc_seq_len) return esc_seq_len;
+
+    bool found = is_color_escape_seq(code, &esc_seq_len);
```

Comment:
No idea. And I doubt the people who wrote this code considered that question. But with this change the question is moot. I've got another change I'll add to this PR that optimizes the cached sequence lookup using feedback from the cache hits so it will be as close to optimal as we can get.